bundles/org.openhab.binding.haywardomnilogic/src/main/java/org/openhab/binding/haywardomnilogic/internal/discovery/HaywardDiscoveryService.java

```diff
@@ -0,0 +1,452 @@
+            DiscoveryResult result = DiscoveryResultBuilder.create(thingUID)
+                    .withBridge(bridgehandler.getThing().getUID())
+                    .withRepresentationProperty(HaywardBindingConstants.PROPERTY_SYSTEM_ID)
+                    .withLabel("Hayward " + label).withProperties(properties).build();
+            thingDiscovered(result);
+        }
+    }
+
+    public void onSensorDiscovered(int systemID, String label, String bowID, String bowName) {
+        logger.trace("Hayward Sensor {} Discovered: {}", systemID, label);
+        HaywardBridgeHandler bridgehandler = discoveryBridgehandler;
+        if (bridgehandler != null) {
+            ThingUID thingUID = new ThingUID(HaywardBindingConstants.THING_TYPE_SENSOR,
+                    bridgehandler.getThing().getUID(), Integer.toString(systemID));
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(HaywardBindingConstants.PROPERTY_SYSTEM_ID, String.valueOf(systemID));
+            properties.put(HaywardBindingConstants.PROPERTY_TYPE, HaywardTypeToRequest.SENSOR);
+            properties.put(HaywardBindingConstants.PROPERTY_BOWID, bowID);
+            properties.put(HaywardBindingConstants.PROPERTY_BOWNAME, bowName);
+            DiscoveryResult result = DiscoveryResultBuilder.create(thingUID)
+                    .withBridge(bridgehandler.getThing().getUID())
+                    .withRepresentationProperty(HaywardBindingConstants.PROPERTY_SYSTEM_ID)
+                    .withLabel("Hayward " + label).withProperties(properties).build();
+            thingDiscovered(result);
+        }
+    }
+
+    public void onVirtualHeaterDiscovered(int systemID, String label, String bowID, String bowName) {
+        logger.trace("Hayward Virtual Heater {} Discovered: {}", systemID, label);
+        HaywardBridgeHandler bridgehandler = discoveryBridgehandler;
+        if (bridgehandler != null) {
+            ThingUID thingUID = new ThingUID(HaywardBindingConstants.THING_TYPE_VIRTUALHEATER,
+                    bridgehandler.getThing().getUID(), Integer.toString(systemID));
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(HaywardBindingConstants.PROPERTY_SYSTEM_ID, String.valueOf(systemID));
+            properties.put(HaywardBindingConstants.PROPERTY_TYPE, HaywardTypeToRequest.VIRTUALHEATER.toString());
+            properties.put(HaywardBindingConstants.PROPERTY_BOWID, bowID);
+            properties.put(HaywardBindingConstants.PROPERTY_BOWNAME, bowName);
+            DiscoveryResult result = DiscoveryResultBuilder.create(thingUID)
+                    .withBridge(bridgehandler.getThing().getUID())
+                    .withRepresentationProperty(HaywardBindingConstants.PROPERTY_SYSTEM_ID)
+                    .withLabel("Hayward " + label).withProperties(properties).build();
+            thingDiscovered(result);
+        }
```

Comment:
Similar code appears in many other places in this class. Please refactor the code into a separate function.